File: Final/Code/Project_4/DFS/Simple_Graph/Simple_Graph.cpp
// dùng đệ quy
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
using namespace std;

void dfs_recursive(const unordered_map<char, vector<char>>& graph, char node, unordered_set<char>& visited) {
    cout << node << " ";
    visited.insert(node);

    for (char neighbor : graph.at(node)) {
        if (visited.find(neighbor) == visited.end()) {
            dfs_recursive(graph, neighbor, visited);
        }
    }
}

int main() {
    unordered_map<char, vector<char>> graph = {
        {'A', {'B', 'C'}},
        {'B', {'A', 'D', 'E'}},
        {'C', {'A', 'F'}},
        {'D', {'B'}},
        {'E', {'B', 'F'}},
        {'F', {'C', 'E'}}
    };

    unordered_set<char> visited;
    cout << "DFS bắt đầu từ đỉnh A:\n";
    dfs_recursive(graph, 'A', visited);

    return 0;
}

File: Final/Code/Project_4/DFS/Simple_Graph/Simple_Graph.py
# dùng đệ quy
def dfs_recursive(graph, start, visited=None):
    if visited is None:
        visited = set()

    print(start, end=' ')
    visited.add(start)

    for neighbor in graph[start]:
        if neighbor not in visited:
            dfs_recursive(graph, neighbor, visited)

graph = {
    'A': ['B', 'C'],
    'B': ['A', 'D', 'E'],
    'C': ['A', 'F'],
    'D': ['B'],
    'E': ['B', 'F'],
    'F': ['C', 'E']
}

print("DFS bắt đầu từ đỉnh A:")
dfs_recursive(graph, 'A')